design-proposals/orch-cli.md

```diff
@@ -0,0 +1,229 @@
+  i.e. `cli update config --endpoint https://my-orchestrator.com/`.
+
+- `completion` ... return a script that can be used to configure autocompletion.
+  TO-DO: Consider treating completion as a verb/noun operation? i.e. `cli get completion bash`.
+
+- `upload <directory>` ... upload items to the Orchestrator. This is currently a catalog-specific
+  operation and uploads a directory of yaml files to the orchestrator.
+
+- `version` ... return the version number of the CLI.
+
+- `watch` ... watch an object in the Orchestrator using webhook or stream interface, returning changes
+  as they occur.
+  TO-DO: drop this?
+
+- `wipe` ... wipe all data in a project.
+  TO-DO: drop this? Should be implemented as part of `cli delete project`.
+
+#### Nouns
+
+This ADR does not seek to exhaustively list the set of nouns that are available on the CLI. Each
+component shall add their own nouns as appropriate. For example,
+
+- `cli list applications` ... The `application` noun applies to applications in the catalog.
+
+- `cli list registry` ... The `registry` noun applies to registries in the catalog.
+
+#### Subjects
+
+Subjects are additional information or context releated to a noun. Different nouns may support
+different subjects. For convenience sake, most nouns support a name subject and optionally for those
+objects that are versioned, a version subject. For example,
+
+- `cli get application nginx` ... returns the first nginx application, agnostic to the version.
+
+- `cli get application nginx 0.0.1` ... return exactly version 0.0.1 of the nginx application.
+
+If additional subjects beyond name and version are required, then they should be specified with addional
+options. For example,
+
+- `cli get application nginx --deployed true` ... get the nginx application that is currently deployed.
+
+#### Global Options
+
+- `-h` / `--help` ... return help. The help returned may be contextualized to the verb or noun that is
+  being used. For example `cli -h` returns global help whereas `cli create application -h` return help
+  that is relevant to creating applications.
+
+- `-n` / `--noauth` ... disables authentication checks, only useful in a devleopment environment.
+
+- `-p` / `--project` ... sets the project that will be used for the current verb. Required for verbs that
+  act on projects.
+
+- `-v` / `--verbose` ... enables verbose output.
+
+- `-o json` / `-o yaml` ... instead of returning human-readable output, return output in `yaml` or `json`.
+
+#### Verb-Specific and/or Noun-Specific options
+
+Verbs and nouns may have addional non-global options as necessary. For example, the `create application`
+verb/noun pair includes the options `--chart-name` and `--chart-registry` and `--chart-version`.
+
+#### Examples
+
+- `cli login sample-project-admin` ... log in as a user.
+
+- `cli -p acme list applications nginx` ... list all nginx applications in a tabular format.
```

Comment:
Does the CLI tooling support using a go template to format output?

Reply:
Not at this time -- I think working on an output formatter would be a good idea, and good to get started on it sooner rather than later, before everyone hand-codes their output code. I'm reminded of the one that was in voltctl, but it's been about 5 years. I'm sure there are better things now.